# Numpy Financial
*Outlines a sleek python package for conducting financial analysis*
[numpy-financial 1.0.0](https://www.example.com)

In [1]:
# lets us run the notebook from within our django project
import django_initializer

## Future value
This first function will allow us to compute the future value for a given asset. This is normally used by investors and financial planners to estimate how much an investment made today will be worth in the future. Using and learning python is great because I am going to show you why investing your time in learning python is way more important than trying to hack all of your processes together in microsoft office products.

In [4]:
import numpy as np
import numpy_financial as npf

What is the future value after 10 years of saving $100 now, with an additional monthly savings of $100? Assume the interest rate is 5% (annually) compounded monthly.

In [5]:
npf.fv(0.05/12, 10*12, -100, -100)

15692.928894335748

By convention, the negative sign represents cash flow out (i.e. money not available today). Thus, saving $100 a month at 5% annual interest leads to $15,692.93 available to spend in 10 years.

Let’s compare different interest rates from the example above. We can pass the function for future value an array that will calculate each value we give it and then return back an array with the answer for each value that was fed. 

In [7]:
"""
Here we are comparing what the future value is for 5%, 6%, or 7% interest rates on the same savings amount of $100 down, then an additional $100 per month invested.
"""
a = np.array((0.05, 0.06, 0.07))/12
npf.fv(a, 10*12, -100, -100)

array([15692.92889434, 16569.87435405, 17509.44688102])

Future value (FV) is the value of a current asset at some point in the future based on an assumed growth rate.
Investors are able to reasonably assume an investment's profit using the future value (FV) calculation.
Determining the future value (FV) of a market investment can be challenging because of the market's volatility.
There are two ways of calculating the future value (FV) of an asset: FV using simple interest and FV using compound interest.

​	  
FV=I×(1+(R×T))

WHERE:

I = Investment amount
R = Interest rate
T = Number of years

FV=I×(1+R^T)

WHERE:

I = Investment amount
R = Interest rate
T = Number of years
​	

## Interest of a given payment

This allows us to calculate amortization and generate a scheduler for paying off a specific loan.


In [8]:
# ------------------------------- #
# imports 
# ------------------------------- #

import numpy as np
import numpy_financial as npf

"""
question: what is the amortization schedule for a 1 year loan of $2500 at 8.24% interest per year compunded?
"""

# setting a principal payment
principal = 2500.00

# setting a period (+ 1 because in finance arrays index at 1)
per = np.arange(1*12) + 1

# ipmt and ppmt are types of payment, interest vs principal
ipmt = npf.ipmt(0.0824/12, per, 1*12, principal)
ppmt = npf.ppmt(0.0824/12, per, 1*12, principal)

# each element of the sum of ipmt and ppmt equal the total payment
pmt = npf.pmt(0.0824/12, 1*12, principal)
np.allclose(ipmt + ppmt, pmt)

True

In [9]:
# string formatting for displaying an amortization schedule
fmt = fmt = '{0:2d} {1:8.2f} {2:8.2f} {3:8.2f}'
for payment in per:
    index = payment - 1
    principal = principal + ppmt[index]
    print(fmt.format(payment, ppmt[index], ipmt[index], principal))

 1  -200.58   -17.17  2299.42
 2  -201.96   -15.79  2097.46
 3  -203.35   -14.40  1894.11
 4  -204.74   -13.01  1689.37
 5  -206.15   -11.60  1483.22
 6  -207.56   -10.18  1275.66
 7  -208.99    -8.76  1066.67
 8  -210.42    -7.32   856.25
 9  -211.87    -5.88   644.38
10  -213.32    -4.42   431.05
11  -214.79    -2.96   216.26
12  -216.26    -1.49    -0.00


In [10]:
# carve off just the interest payments
interestpd = np.sum(ipmt)
np.round(interestpd, 2)

-112.98